packages/concerto-tools/lib/codegen/fromopenapi/cto/concertometamodelvisitor.js

```diff
@@ -0,0 +1,522 @@
+// }
+
+function flattenOaiAnyOfOrOneOf([oaiPropertyName, oaiPropertyBody]) {
+    const alternative = oaiPropertyBody.anyOf ||
+        oaiPropertyBody.oneOf;
+
+    if (alternative) {
+        const keyword = oaiPropertyBody.anyOf ? 'anyOf' : 'oneOf';
+        console.warn(
+            `Keyword '${keyword}' in definition '${oaiPropertyName}' is not fully supported. Defaulting to first alternative.`
+        );
+
+        delete oaiPropertyBody.anyOf;
+        delete oaiPropertyBody.oneOf;
+
+        return [
+            oaiPropertyName, {
+                ...oaiPropertyBody,
+                ...alternative
+            }
+        ];
+    }
+
+    return [oaiPropertyName, oaiPropertyBody];
+}
+
+function isOaiPropertyUnmodelledJson(oaiPropertyBody) {
+    return oaiPropertyBody.type === 'object' &&
+        typeof oaiPropertyBody.additionalProperties === 'object';
+}
+
+function oaiTypeToConerto$class([oaiPropertyName, oaiPropertyBody]) {
+    if (isOaiPropertyArray(oaiPropertyBody)) {
+        return oaiTypeToConertoBasic$class([
+            oaiPropertyName,
+            oaiPropertyBody.items
+        ]);
+    }
+
+    if (isOaiPropertyObject(oaiPropertyBody)) {
+        return;
+    }
+
+    return oaiTypeToConertoBasic$class([oaiPropertyName, oaiPropertyBody]);
+}
+
+function oaiToConcertoProperty([oaiPropertyName, oaiPropertyBody]) {
+    // const oaiPropertyWithout
+    return {
+        $class: oaiTypeToConerto$class([oaiPropertyName, oaiPropertyBody]),
+        isArray: isOaiPropertyArray(oaiPropertyBody),
+        isOptional: false // TODO:
+    };
+}
+
+function oaiToConcertoProperties(oaiProperties) {
+    return Object.entries(oaiProperties)
+        .map(flattenOaiAnyOfOrOneOf)
+        // @ts-ignore
+        .map(oaiToConcertoProperty);
+}
+
+function oaiObjectToDeclaration([oaiObjectK, oaiObjectV]) {
+    const properties = oaiToConcertoProperties(oaiObjectV.properties);
+    return {
+        $class: `${parameters.metaModelNamespace}.ConceptDeclaration`,
+        name: oaiObjectK,
+        isAbstract: false,
+        properties
+    };
+}
+
+
+class Visitable {
+    accept(visitor, parameters) {
+        return visitor.visit(this, parameters);
+    }
+}
+
+class BodyVisitable extends Visitable {
+    constructor(body) {
+        super();
+        this.body = body;
+    }
+}
+
+class NameAndBodyVisitable extends Visitable {
+    constructor(name, body) {
+        super();
+        this.name = name;
+        this.body = body;
+    }
+}
+
+class OaiLocalReference extends BodyVisitable {
+    isOaiLocalReference = true;
+}
+
+class OaiReference extends BodyVisitable {
+    isOaiReference = true;
+}
+
+class OaiProperty extends NameAndBodyVisitable {
+    isOaiProperty = true;
+}
+
+class OaiProperties extends BodyVisitable {
+    isOaiProperties = true;
+}
+
+class OaiSchema extends NameAndBodyVisitable {
+    isOaiSchema = true;
+}
+
+class OaiSchemas extends BodyVisitable {
+    isOaiSchemas = true;
+}
+
+class OpenApiVisitor {
```

Comment:
Please be consistent with the filename and default export

Reply:
QQ: Is "CTO" strictly used for the Concerto format, or can it be used to refer to the Concerto metamodel JSON as well. If the former, what would you advise that the best to refer to Concerto metamodel JSON be:
1. `concertoMetamodelJson`
2. `ctoMetamodelJson`
3. `ctoMtmdlJson`
4. `concertoMetamodel`
5. `ctoMetamodel`
6. `ctoMtmdl`
7. `cto`
8. `metamodel`

I'm aiming to find a way to refer to the output, in succinct way, while avoiding confusion with non-JSON Concerto files and the base Concerto metamodel.

Reply:
Yes, generally speaking, we use `CTO` to mean the Concerto Modelling Language (CML) format.

Reply:
Addressed ✅